contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,253 @@
+to wrap `kubectl` so the tool is invoked on every command. For example, the following
+is a [real example](
+https://github.com/heptio/authenticator#4-set-up-kubectl-to-use-heptio-authenticator-for-aws-tokens)
+from Heptio's AWS authenticator:
+
+```
+kubectl --kubeconfig /path/to/kubeconfig --token "$(heptio-authenticator-aws token -i CLUSTER_ID -r ROLE_ARN)" [...]
+```
+
+Beside resulting in a long command, this potentially encourages distributions to
+wrap or fork kubectl, changing the way that users interact with different
+Kubernetes clusters.
+
+# Proposal
+
+This proposal builds off of earlier requests to [support exec-based plugins](
+https://github.com/kubernetes/kubernetes/issues/35530#issuecomment-256170024), and
+proposes that we should add this as a first-class feature of kubectl. Specifically,
+client-go should be able to receive credentials by executing a command and reading
+that command's stdout.
+
+In fact, client-go already does this today. The GCP plugin can already be configured
+to [call a command](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.8.5/plugin/pkg/client/auth/gcp/gcp.go#L228-L240)
+other than `gcloud`.
+
+## Plugin responsibilities
+
+Plugins are exec'd through client-go and print credentials to stdout. Errors are
+surfaced through stderr and a non-zero exit code. To use the bearer token
+`BEARER_TOKEN`, a plugin would print:
+
+```terminal
+$ ./kubectl-example-auth-plugin
+BEARER_TOKEN
+```
+
+Plugin output doesn't use versioned Kubernetes APIs (with kind and apiVersion)
+because we expect this plugin to interact with tools that don't natively understand
+Kubernetes. For example, the GCP auth provider execs [`gcloud auth`](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.9.0/plugin/pkg/client/auth/gcp/gcp.go#L230)
+and parses the output using JSONPath.
+
+To surface runtime-based information to the plugin, client-go will set environment
+variables prefixed with the string `KUBERNETES_`, similar to [CNI](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md#parameters).
+This will be used to support context based credential strategies like [request signing](https://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html),
+and allow plugins to respond to events such as 401 HTTP status codes.
+
+This document does NOT propose any specific environment variables, but some examples
+might include:
+
+```
+KUBERNETES_VERSION="v1.10.4"
+KUBERNETES_REQUEST_BODY='{"apiVersion", ... }'
+KUBERNETES_LAST_STATUS_CODE="401"
+```
+
+### Caching
+
+kubectl repeatedly [re-initializes transports](https://github.com/kubernetes/kubernetes/issues/37876)
+while client-go transports are long lived over many requests. As a result naive auth
+provider implementations that re-request credentials on every request have historically
+been slow.
+
+Initially, plugins will be responsible for caching credentials and knowing when those
```

Comment:
If we get a 401, is it ever unsafe to retry to with new credentials?  I can't think of a case where that would break.

I'm trying to think of why (other than ease of implementation), we would choose to call on every request instead of on first transport and (maybe) on 401s.  I'm concerned about a plugin counting on being called every request for some reason.

Reply:
> I'm trying to think of why (other than ease of implementation), we would choose to call on every request instead of on first transport and (maybe) on 401s. I'm concerned about a plugin counting on being called every request for some reason.

This actually sounds pretty reasonable behavior to start with.

Reply:
So @deads2k you are recommending that the plugin would return a token and an expiry, and would only be executed if client-go had no cached token, if it was expired, or if client-go received a 401 in which case it would get another token and retry?

Reply:
> So @deads2k you are recommending that the plugin would return a token and an expiry, and would only be executed if client-go had no cached token, if it was expired, or if client-go received a 401 in which case it would get another token and retry?

I'm suggesting that we could ask for a token the first time we make the transport and after that we call out on 401 responses only instead on every call.  It does a couple things: first it prevents the plugin from getting used to being called every request and second, it allows us to only make the external process calls when we need to.